cloud/scope/cluster.go

```diff
@@ -105,16 +107,32 @@ func (s *ClusterScope) Network() *infrav1.Network {
 }
 
 // PublicIPSpec returns the public IP specs.
-func (s *ClusterScope) PublicIPSpecs() []azure.PublicIPSpec {
-	return []azure.PublicIPSpec{
+func (s *ClusterScope) PublicIPSpecs() ([]azure.PublicIPSpec, error) {
+	spec := []azure.PublicIPSpec{
 		{
 			Name: azure.GenerateNodeOutboundIPName(s.ClusterName()),
 		},
-		{
-			Name:    s.Network().APIServerIP.Name,
-			DNSName: s.Network().APIServerIP.DNSName,
-		},
 	}
+
+	// If IP provided by name, expect to exist and don't precreate.
+	if s.AzureCluster.Spec.NetworkSpec.PublicIP != nil && s.AzureCluster.Spec.NetworkSpec.PublicIP.Name != "" {
+		return spec, nil
+	}
+
+	if s.Network().APIServerIP.Name == "" {
+		h := fnv.New32a()
+		if _, err := h.Write([]byte(fmt.Sprintf("%s/%s/%s", s.SubscriptionID(), s.ResourceGroup(), s.ClusterName()))); err != nil {
+			return nil, errors.Wrapf(err, "failed to write hash sum for api server ip")
+		}
+		s.Network().APIServerIP.Name = azure.GeneratePublicIPName(s.ClusterName(), fmt.Sprintf("%x", h.Sum32()))
```

Comment:
It seems odd to be setting the value in here given that it's implied to be a Getter function, especially since this is setting status... We do need this value to be set before reconcile IPs is called. Thoughts about breaking into a separate function calling it something explicit like `SetAPIServerIP()` and calling it from the reconciler like we do with `SetFailureDomains()`? 

I realize that's a partial reversal to what you had before so I apologize if my initial comment was confusing.

Reply:
I would opt toward a setter as it would reduce the depth of references and be specific about what the method requires from its scope interface.

Reply:
better?